### ISSUE TITLE
[false-negative]DMI_INVOKING_TOSTRING_ON_ARRAY can not detect String automatic conversion when string concatenation

### ISSUE THREAD

**Issue by wuchiuwong** [spotbugs/spotbugs#1147]
**Affects SpotBugs Version:** 
spotbugs-4.0.0-RC2

**Rule:**
DMI_INVOKING_TOSTRING_ON_ARRAY

**Description:**
Unable to detect String automatic conversion when string concatenation, for example: <code>"foo" + barArray</code>,
This rule is implemented in SpotBugs.jar\main\java\edu\umd\cs\findbugs\detect\FindPuzzlers.java.
This rule uses the following code to detect explicit string conversion:
```
if (seen == Const.INVOKEVIRTUAL
                && stack.getStackDepth() > 0
                && ("toString".equals(getNameConstantOperand()) && "()Ljava/lang/String;".equals(getSigConstantOperand())
                        || "append".equals(getNameConstantOperand())
                                && "(Ljava/lang/Object;)Ljava/lang/StringBuilder;".equals(getSigConstantOperand())
                                && "java/lang/StringBuilder".equals(getClassConstantOperand())
                        || "append".equals(getNameConstantOperand())
                                && "(Ljava/lang/Object;)Ljava/lang/StringBuffer;".equals(getSigConstantOperand())
                                && "java/lang/StringBuffer".equals(getClassConstantOperand()) || ("print".equals(getNameConstantOperand())
                                        || "println".equals(getNameConstantOperand()))
                                        && "(Ljava/lang/Object;)V".equals(getSigConstantOperand())))
```
which includes the following 5 situations: 1.<code>toString</code>, 2.<code>StringBuilder.append(Object)</code>, 3.<code>StringBuffer.append(Object)</code>, 4.<code>print(Object)</code>, 5.<code>println(Object)</code>.
The string conversion performed during string concatenation is ignored.

**Code Sample demonstrating the issue:**
```
public void setRetrieveAttributes(String[] retrieveAttributes)
{
	LOGGER.debug("retrieveAttributes::" + retrieveAttributes);
}
```

**Expected outcome:**
false-negative

**Running SpotBugs through:**
command line

**Comment by gloNelson** [comment]
@wuchiuwong Sorry, I cannot reproduce the issue.

SpotBugs Version:
spotbugs-4.0.0-RC2
Java Version:
openjdk1.8 (Also jdk 11)

Bytecode file:
```// class version 52.0 (52)
// access flags 0x21
public class Issue1147 {

  // compiled from: Issue1147.java

  // access flags 0x1
  public <init>()V
   L0
    LINENUMBER 3 L0
    ALOAD 0
    INVOKESPECIAL java/lang/Object.<init> ()V
    RETURN
   L1
    LOCALVARIABLE this LghFalseNegative/Issue1147; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1

  // access flags 0x1
  public setRetrieveAttributes([Ljava/lang/String;)V
   L0
    LINENUMBER 6 L0
    GETSTATIC java/lang/System.out : Ljava/io/PrintStream;
    NEW java/lang/StringBuilder
    DUP
    INVOKESPECIAL java/lang/StringBuilder.<init> ()V
    LDC "retrieveAttributes::"
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/String;)Ljava/lang/StringBuilder;
    ALOAD 1
    INVOKEVIRTUAL java/lang/StringBuilder.append (Ljava/lang/Object;)Ljava/lang/StringBuilder;
    INVOKEVIRTUAL java/lang/StringBuilder.toString ()Ljava/lang/String;
    INVOKEVIRTUAL java/io/PrintStream.println (Ljava/lang/String;)V
   L1
    LINENUMBER 7 L1
    RETURN
   L2
    LOCALVARIABLE this LghFalseNegative/Issue1147; L0 L2 0
    LOCALVARIABLE retrieveAttributes [Ljava/lang/String; L0 L2 1
    MAXSTACK = 3
    MAXLOCALS = 2
}
```
As you can see, the string concatenation optimizes by using the StringBuilder class. In this case, the situation can satisfy the code you showed earlier.

Can you tell me your Java version? Thanks a lot.

**Comment by pzygielo** [comment]
I observe the same false-negative.

For
```java
package pzrep;

import javax.transaction.xa.Xid;

class ToStringOnArray {
  String convert1(Xid id) {
    return "Prefix " + id.getGlobalTransactionId() + " Suffix";
  }
}
```
1. javac 8 generates:
```
Compiled from "ToStringOnArray.java"
class pzrep.ToStringOnArray {
  pzrep.ToStringOnArray();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  java.lang.String convert1(javax.transaction.xa.Xid);
    Code:
       0: new           #2                  // class java/lang/StringBuilder
       3: dup
       4: invokespecial #3                  // Method java/lang/StringBuilder."<init>":()V
       7: ldc           #4                  // String Prefix
       9: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      12: aload_1
      13: invokeinterface #6,  1            // InterfaceMethod javax/transaction/xa/Xid.getGlobalTransactionId:()[B
      18: invokevirtual #7                  // Method java/lang/StringBuilder.append:(Ljava/lang/Object;)Ljava/lang/StringBuilder;
      21: ldc           #8                  // String  Suffix
      23: invokevirtual #5                  // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      26: invokevirtual #9                  // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      29: areturn
}
```
which *IS* recognized by spotbugs (4.3.0) and violation is reported:
```
H C USELESS_STRING: Invocation of toString on javax.transaction.xa.Xid.getGlobalTransactionId() in pzrep.ToStringOnArray.convert1(Xid)  At ToStringOnArray.java:[line 7]
```

2. but with javac 11 (also checked with 16) - `makeConcatWithConstants` is used:
```
Compiled from "ToStringOnArray.java"
class pzrep.ToStringOnArray {
  pzrep.ToStringOnArray();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  java.lang.String convert1(javax.transaction.xa.Xid);
    Code:
       0: aload_1
       1: invokeinterface #2,  1            // InterfaceMethod javax/transaction/xa/Xid.getGlobalTransactionId:()[B
       6: invokedynamic #3,  0              // InvokeDynamic #0:makeConcatWithConstants:([B)Ljava/lang/String;
      11: areturn
}
```
and the same bug **is not recognized**.

**Comment by swpalmer** [comment]
Is this related to this message I get running on recent JDKs:
    The following classes needed for analysis were missing:
      java.lang.invoke.StringConcatFactory

?
